### DISCUSSION TITLE
GitHub GraphQL API not getting recent contributions from other users.

### DISCUSSION THREAD

**Discussion by MichaelSimpsonVertex** [community/community#24800] [Apps, API and Webhooks]
<p>I am trying to get the most recent contributions from the <code>ContributionsCollection</code> return from any user in our organization. I have been using <a href="https://developer.github.com/v4/explorer/" rel="nofollow noopener">https://developer.github.com/v4/explorer/</a> for testing the queries. </p>
<p>How can I see the rest of personB’s (second test) repositories? </p>
<p>I simplified the query to be one user for testing.</p>
<p>When I use my account:</p>
<pre><code>query {
 user(login: "MichaelSimpsonVertex") {
    login,
    contributionsCollection(organizationID: "MDEyOk9yZ2FuaXphdGlvbjgxNzcyNw==", to:"2019-04-01T01:01:00", from: "2019-01-01T01:01:00") {
      startedAt,
      endedAt,
      hasAnyContributions,
      hasActivityInThePast,
      commitContributionsByRepository {
        repository{
          name
        },
        contributions (first:1, orderBy:{field: OCCURRED_AT, direction: DESC}) {
          nodes{
            occurredAt,
            commitCount
          }
        }
      }
    }
  }
}
</code></pre>
<p>Output show multiple repositories (I replaced the repository names):</p>
<pre><code>{
  "data": {
    "user": {
      "login": "MichaelSimpsonVertex",
      "contributionsCollection": {
        "startedAt": "2019-01-01T09:01:00Z",
        "endedAt": "2019-04-01T08:01:00Z",
        "hasAnyContributions": true,
        "hasActivityInThePast": true,
        "commitContributionsByRepository": [
          {
            "repository": {
              "name": "madeupname1"
            },
            "contributions": {
              "nodes": [
                {
                  "occurredAt": "2019-02-04T08:00:00Z",
                  "commitCount": 3
                }
              ]
            }
          },
          {
            "repository": {
              "name": "madeupname2"
            },
            "contributions": {
              "nodes": [
                {
                  "occurredAt": "2019-02-11T08:00:00Z",
                  "commitCount": 6
                }
              ]
            }
          },
          {
            "repository": {
              "name": "madeupname3"
            },
            "contributions": {
              "nodes": [
                {
                  "occurredAt": "2019-03-27T07:00:00Z",
                  "commitCount": 3
                }
              ]
            }
          },
          {
            "repository": {
              "name": "madeupname4"
            },
            "contributions": {
              "nodes": [
                {
                  "occurredAt": "2019-03-07T08:00:00Z",
                  "commitCount": 3
                }
              ]
            }
          },
          {
            "repository": {
              "name": "madeupname5"
            },
            "contributions": {
              "nodes": [
                {
                  "occurredAt": "2019-01-14T08:00:00Z",
                  "commitCount": 2
                }
              ]
            }
          },
          {
            "repository": {
              "name": "madeupname6"
            },
            "contributions": {
              "nodes": [
                {
                  "occurredAt": "2019-01-04T08:00:00Z",
                  "commitCount": 7
                }
              ]
            }
          },
          {
            "repository": {
              "name": "madeupname7"
            },
            "contributions": {
              "nodes": [
                {
                  "occurredAt": "2019-03-14T07:00:00Z",
                  "commitCount": 1
                }
              ]
            }
          },
          {
            "repository": {
              "name": "madeupname8"
            },
            "contributions": {
              "nodes": [
                {
                  "occurredAt": "2019-03-14T07:00:00Z",
                  "commitCount": 2
                }
              ]
            }
          },
          {
            "repository": {
              "name": "madeupname9"
            },
            "contributions": {
              "nodes": [
                {
                  "occurredAt": "2019-03-15T07:00:00Z",
                  "commitCount": 1
                }
              ]
            }
          }
        ]
      }
    }
  }
}
</code></pre>
<p>When I try someone elses account (replaced user name after running test)</p>
<pre><code>query {
   user(login: "personB") {
    login,
    contributionsCollection(organizationID: "MDEyOk9yZ2FuaXphdGlvbjgxNzcyNw==", to:"2019-04-01T01:01:00", from: "2019-01-01T01:01:00") {
      startedAt,
      endedAt,
      hasAnyContributions,
      hasActivityInThePast,
      commitContributionsByRepository {
        repository{
          name
        },
        contributions (first:1, orderBy:{field: OCCURRED_AT, direction: DESC}) {
          nodes{
            occurredAt,
            commitCount
          }
        }
      }
    }
  }
}
</code></pre>
<p>Output:</p>
<pre><code>{
  "data": {
    "user": {
      "login": "personB",
      "contributionsCollection": {
        "startedAt": "2019-01-01T09:01:00Z",
        "endedAt": "2019-04-01T08:01:00Z",
        "hasAnyContributions": true,
        "hasActivityInThePast": true,
        "commitContributionsByRepository": [
          {
            "repository": {
              "name": "madeupname6"
            },
            "contributions": {
              "nodes": [
                {
                  "occurredAt": "2019-03-20T07:00:00Z",
                  "commitCount": 1
                }
              ]
            }
          }
        ]
      }
    }
  }
}
</code></pre>
<p>Thanks</p>

**Comment by lee-dohm** [comment]
<p>I’m not sure I understand what the problem is here. Can you describe what the problem is with the output that you’re seeing in the error case?</p>

**Comment by clayallsopp** [comment]
<p><a class="mention" href="/lee-dohm">@lee-dohm</a> <a class="mention" href="/michaelsimpsonvertex">@michaelsimpsonvertex</a> I believe I’m seeing something similar - the specific problem is when I view another user’s profile via HTML (i.e. <a href="http://github.com/foo-user" class="inline-onebox-loading" rel="nofollow noopener">github.com/foo-user</a>) I am seeing many more contributions than when accessed via GraphQL similar to what <a class="mention" href="/michaelsimpsonvertex">@michaelsimpsonvertex</a> posted. The contributions look correct when I view my own data.</p>

**Comment by clayallsopp** [comment, marked as answer]
<p>Upon further investigation, it looks like the GraphQL version is only returning the “publicly viewable” contribution metrics - it does not seem to take into “viewer” account’s permissions</p>

**Comment by MichaelSimpsonVertex** [comment]
<p>Thanks for the response,  your are right as far as I know. I forgot to put my findings into this thread.</p>
<p>I would expect a user with organization admin to be able to get all activity from repositories the organization has access to.</p>
<p>But I believe since the GraphQL object is only showing ojects from the User and not the Organization perspective, the response is limting access to the user who own the token. So using your token you can only see public repo activity from other users, but can see all repo activity from yourself.</p>
<p>I have put in a feature request, via email, to get a organizationUserActivty object into graphQL that will hopefully supply a attribute for “latestUserActivtyDate” Or at least the specific list of activity done to repositoires in the organization.</p>
<p>Thanks</p>

**Comment by ilia-cy** [comment]
<p>Hey <a class="mention" href="/michaelsimpsonvertex">@michaelsimpsonvertex</a> ,</p>
<p>Did you get a response for the feature request?</p>
<p>Is it documented somewhere in which we can see its status?</p>
<p>Thanks</p>

**Comment by MichaelSimpsonVertex** [comment]
<p>I started an email exchange with GitHub Developer Support. In short, I was not able to find a simple solution for this problem. Below is what I found out.</p>
<p><strong>My question:</strong></p>
<blockquote>
<p>I am trying to get the last activity a member that has done on any organization’s repository. This would involve Issues updates, Commits, Pulls to local, or Merges, done to repositories that the organization owns. I want the each users in the organization latest activity in an organization in date form.</p>
</blockquote>
<blockquote>
<p> </p>
</blockquote>
<blockquote>
<p>The Information I get will be used to remove members from the organization if they have been inactive 90 days or more.</p>
</blockquote>
<p><strong>Answers I recieved</strong></p>
<blockquote>
<p>Matt Barber email</p>
</blockquote>
<blockquote>
<p>I think what you are looking for is a way of collecting these events as they happen over a rolling 90 day window.</p>
</blockquote>
<blockquote>
<p>For scenarios such as this we provide webhooks: <a href="https://developer.github.com/webhooks/" rel="nofollow noopener">https://developer.github.com/webhooks/</a> webhooks</p>
</blockquote>
<blockquote>
<p>Webhooks can be installed on an organization or a specific repository. Once installed, the webhook will be triggered each time one or more subscribed events occurs, sending a payload representing that event to an endpoint on your server, that is configured with your hook.</p>
</blockquote>
<blockquote>
<p>So, for instance, if you wanted to look at issue events:</p>
</blockquote>
<blockquote>
<p><a href="https://developer.github.com/v3/activity/events/types/#issuesevent" rel="nofollow noopener">https://developer.github.com/v3/activity/events/types/#issuesevent</a></p>
</blockquote>
<blockquote>
<p>These are triggered whenever an issue is opened, edited, deleted, transferred, pinned, unpinned, closed, reopened, assigned, unassigned, labeled, unlabeled, milestoned, or demilestoned.</p>
</blockquote>
<blockquote>
<p>In the payload the sender is defined, and this is the user that performed the action. As these are close to realtime events, you can calculate that the time you receive this is the time the action happened.</p>
</blockquote>
<blockquote>
<p>The caveat with this is that it doesn’t provide back data, so once you configure - from that moment forwards you will be able to begin using these payloads.</p>
</blockquote>
<blockquote>
<p>I hope this helps, please let me know how you get on, and if you face any issues or have any questions while looking into this.</p>
</blockquote>
<blockquote>
<p>Thanks<br>
Matt Barber</p>
</blockquote>
<blockquote>
<p>Ivan Zuzak email</p>
</blockquote>
<blockquote>
<p> </p>
</blockquote>
<blockquote>
<p>Some of the the information you’re looking for (e.g. pulls to a local clone) are not available via the API or webhooks – there’s really no way to get such data currently. Some other information you mentioned is available only via webhooks, and some data is available via the API. The data that’s available via the API is scattered over multiple API endpoints, similar to the UI, so you’d need to collect all that data with multiple API requests.</p>
</blockquote>
<blockquote>
<p>The team wrote a script which uses one possible definition of activity/inactivity and makes such API calls to determine inactive users:</p>
</blockquote>
<blockquote>
<p><a href="https://github.com/github/platform-samples/tree/master/api/ruby/find-inactive-members" rel="nofollow noopener">https://github.com/github/platform-samples/tree/master/api/ruby/find-inactive-members</a></p>
</blockquote>
<blockquote>
<p>Perhaps that helps you write a script of your own based on your own definition of activity/inactivity.</p>
</blockquote>
<blockquote>
<p>I agree it would be helpful if GitHub itself had a definition of inactivity and if that data was easily available for you to inspect and act upon. I’d be happy to pass this along to the team to consider for the future, but can’t promise when something like that might happen.</p>
</blockquote>
<blockquote>
<p>If you have any other questions – please let us know.</p>
</blockquote>
<blockquote>
<p>Best,<br>
Ivan</p>
</blockquote>
